src/topology_controller.rs

```diff
@@ -0,0 +1,232 @@
+            info!(
+                "ConfigBuilder transforms {:?}",
+                config_builder_new.transforms
+            );
+            info!("ConfigBuilder sinks {:?}", config_builder_new.sinks);
+            topology
+                .reload_config_and_respawn(config_builder_new.build().unwrap())
+                .await
+                .unwrap();
+        }
+        ConfigAction::DELETE => {
+            // source and transform can not use same name in vector
+            let mut config_builder_new = config_builder.clone();
+
+            for id in &config_event.config_ids {
+                let key = &ComponentKey::from(&id);
+                if config_builder_new.sources.get(key).is_some() {
+                    config_builder_new.sources.remove(key);
+                } else if config_builder_new.transforms.get(key).is_some() {
+                    config_builder_new.transforms.remove(key);
+                }
+            }
+            info!("ConfigBuilder sources {:?}", config_builder_new.sources);
+            info!(
+                "ConfigBuilder transforms {:?}",
+                config_builder_new.transforms
+            );
+            info!("ConfigBuilder sinks {:?}", config_builder_new.sinks);
+            topology
+                .reload_config_and_respawn(config_builder_new.build().unwrap())
+                .await
+                .unwrap();
+        }
+        ConfigAction::EXIT => {
+            // should not go here
+        }
+    }
+}
+
+impl TopologyController {
+    pub fn new() -> Self {
+        Self {
+            config_message_sender: None,
+            vector_thread_join_handle: None,
+            generation_id: AtomicU32::new(0),
+        }
+    }
+
+    // start a new thread to run a topology
+    pub fn start(&mut self, topology_config: &str) -> Result<bool, &'static str> {
+        START.call_once(|| {
+            setup_logging();
+        });
+
+        info!("start vector service");
+
+        let res = format::deserialize(topology_config, Some(config::Format::Json));
+        if res.is_err() {
+            error!("deserialize error {:?}", res.unwrap_err());
+            return Err("failed to deserialize config string");
+        }
+        let mut config_builder: ConfigBuilder = res.unwrap();
+        debug!(
+            "config_builder deserialized; sources={:?} transforms={:?} sinks={:?} global={:?}",
+            config_builder.sources,
+            config_builder.transforms,
+            config_builder.sinks,
+            config_builder.global
+        );
+
+        let mut config_builder_copy = config_builder.clone();
+        let builder_for_schema = config_builder.clone();
+        let init_result = config::init_log_schema_from_builder(builder_for_schema, false);
+
+        let (config_message_sender, mut config_message_receiver) = tokio::sync::mpsc::channel(1);
+        self.config_message_sender = Some(config_message_sender);
```

Comment:
The channel sender/receiver is setup here. The sender is copiable and saved into the controller, and the receiver is moved into the loop below (one receiver only). So there is no need to use global variable and unsafe code.